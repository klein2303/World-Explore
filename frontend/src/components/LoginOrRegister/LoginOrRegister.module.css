.loginOrRegister {
    margin: 100px auto;
    width: 500px;
    height: auto;
    padding: 20px;
    display: flex;
    flex-direction: column;
    align-items: center;
}

.title h3 {
    text-align: center;
    font-size: 24px;
    color: #333;
    font-weight: 300;
    margin-bottom: 10px;
}

.title h4  {
    font-weight: 200;
}

.line {
    border: 1px solid #ccc;
    width: 100%;
    margin: 20px 0;
}

.inputSections h5  {
    color: #000000;
    font-weight: 300;
}

.inputSections {
    margin-top: 10px;
    width: 100%;
    display: flex;
    flex-direction: column;
    gap: 15px;
    align-items: center;
    text-align: center;
}

.inputField {
    width: 100%;
    max-width: 300px;
    padding: 10px;
    box-sizing: border-box;
    border: 1px solid #ccc;
    border-radius: 10px;
    font-size: 16px;
}

.inputField::placeholder {
    color: #b5a8a8;
    color: #b5a8a8;
}

.submitButton {
    width: 100%;
    max-width: 300px;
    padding: 10px 20px;
    background-color: #424242;
    border: none;
    border-radius: 15px;
    cursor: pointer;
    font-size: 16px;
    color: #e4e1f7;
    color: #e4e1f7;
}

.submitButton:hover {
    background-color: #351939;
}

@media (max-width: 1030px) {
    .loginOrRegister {
        width: 85%;
        max-width: none;
        padding: 5vw;
        height: 100%;
    }

    .inputField,
    .submitButton {
        width: 90%;
        max-width: none;
        /* Ensure fields also scale with the container */
    }
}

@media (width: 1024px) {
    .loginOrRegister {
        width: 85%;
        max-width: none;
        /* Remove the fixed max-width for better scaling */
        padding: 5vw;
    }

    .inputField,
    .submitButton {
        width: 90%;
        max-width: none;
        /* Ensure fields also scale with the container */
    }
}

