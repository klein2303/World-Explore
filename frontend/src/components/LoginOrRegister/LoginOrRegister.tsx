import {useState} from "react";
import styles from "./LoginOrRegister.module.css";

interface Profile {
    name: string;
    email: string;
    password: string;
}

const LoginOrRegister = () => {
    const [isLogin, setIsLogin] = useState<boolean>(false);
    const [formValues, setFormValues] = useState<Profile>({
        name: "",
        email: "",
        password: "",
    });
    const [isValidPassword, setIsValidPassword] = useState<boolean>(false);
    const [feedbackMessage, setFeedbackMessage] = useState<string>("Password must at least be 8 characters");
    // Load existing users from local storage
    const loadUsers = (): Profile[] => {
        const storedUsers = localStorage.getItem("userProfiles");
        return storedUsers ? JSON.parse(storedUsers) : [];
    };
    const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        const {name, value} = event.target;
        setFormValues({
            ...formValues,
            [name]: value,
        });
        // Password validation: Check if password has at least 8 characters
        if (name === "password") {
            setIsValidPassword(value.length >= 8);
        }
    };

    const handleSubmit = () => {
        const existingUsers = loadUsers();
        if (formValues.name.length === 0 || formValues.email.length === 0 || formValues.password.length === 0) {
            setFeedbackMessage("Please enter all of the fields");
        } else if (existingUsers.some((user) => user.email === formValues.email)) {
            setFeedbackMessage("Email already exists. Please use a different email.");
        } else if (!isValidPassword) {
            setFeedbackMessage("Password is not valid. Password must have 8 characters.");
        } else {
            existingUsers.push(formValues);
            localStorage.setItem("userProfiles", JSON.stringify(existingUsers));
            setFormValues({
                name: "",
                email: "",
                password: "",
            });
            setIsValidPassword(false);
            setFeedbackMessage("Account is made successfully");

            // Delay the switch to login form by 1 second
            setTimeout(() => {
                setIsLogin(true);
            }, 1000);
        }
    };

    return (
        <section aria-label="Login og register component">
            {isLogin ? (
                <section>
                    {/*Make login component here*/}
                    <title aria-label="Login title">Login</title>
                    <h4>User</h4>
                </section>
            ) : (
                <section className={styles.loginOrRegister} aria-label="Register a new account">
                    {/*Register component*/}
                    <div className={styles.title} aria-label="Title">
                        <h3 aria-label="Create a new account">Create an account</h3>
                    </div>
                    <hr className={styles.line} />
                    <section className={styles.inputSections} aria-label="Input fields">
                        <input
                            className={styles.inputField}
                            type="text"
                            name="name"
                            value={formValues.name}
                            onChange={handleInputChange}
                            placeholder="Place enter your name"
                            aria-label="Name input field"
                        />
                        <input
                            className={styles.inputField}
                            type="email"
                            name="email"
                            value={formValues.email}
                            onChange={handleInputChange}
                            placeholder="Please enter your email"
                            aria-label="Email input field"
                        />
                        <input
                            className={styles.inputField}
                            type="password"
                            name="password"
                            value={formValues.password}
                            onChange={handleInputChange}
                            placeholder="Please enter a password"
                            aria-label="Password input field"
                        />
                        <h5 aria-label={feedbackMessage}> {feedbackMessage}</h5>
                        <button
                            className={styles.submitButton}
                            onClick={handleSubmit}
                            aria-label="Submit to create a new account">
                            Create account
                        </button>
                    </section>
                </section>
            )}
        </section>
    );
};

export default LoginOrRegister;
