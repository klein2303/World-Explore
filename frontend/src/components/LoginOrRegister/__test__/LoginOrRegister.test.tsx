import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import { describe, it, expect, beforeEach, vi } from "vitest";
import { MemoryRouter } from "react-router-dom";
import LoginOrRegister from "../LoginOrRegister";

describe("LoginOrRegister Component - Register tests", () => {
    beforeEach(() => {
        // Mock localStorage before each test to prevent actual browser interaction
        localStorage.clear();
        vi.spyOn(Storage.prototype, "getItem").mockReturnValue(null);
        vi.spyOn(Storage.prototype, "setItem").mockImplementation(() => {});
    });

    it("renders the registration form correctly", () => {
        const { asFragment } = render(
            <MemoryRouter>
                <LoginOrRegister login={false} />
            </MemoryRouter>,
        );

        // Snapshot the initial render
        expect(asFragment()).toMatchSnapshot();

        // Check if the form is rendered correctly with input fields and button
        expect(screen.getByLabelText("Name input field")).toBeInTheDocument();
        expect(screen.getByLabelText("Email input field")).toBeInTheDocument();
        expect(screen.getByLabelText("Password input field")).toBeInTheDocument();
        expect(screen.getByLabelText("Submit to create a new account")).toBeInTheDocument();
    });

    it("validates form and handles submission", () => {
        render(
            <MemoryRouter>
                <LoginOrRegister login={false} />
            </MemoryRouter>,
        );

        // Fill out the form with valid values
        fireEvent.change(screen.getByLabelText("Name input field"), { target: { value: "John Doe" } });
        fireEvent.change(screen.getByLabelText("Email input field"), { target: { value: "john@example.com" } });
        fireEvent.change(screen.getByLabelText("Password input field"), { target: { value: "12345678" } });

        // Submit the form
        fireEvent.click(screen.getByLabelText("Submit to create a new account"));

        // Check if the feedback message updates to successful registration
        expect(screen.getByText("Account is made successfully")).toBeInTheDocument();
    });

    it("validates form and handles submission", async () => {
        render(
            <MemoryRouter>
                <LoginOrRegister login={false} />
            </MemoryRouter>,
        );

        // Fill out the form
        fireEvent.change(screen.getByLabelText("Name input field"), { target: { value: "John Doe" } });
        fireEvent.change(screen.getByLabelText("Email input field"), { target: { value: "john@example.com" } });
        fireEvent.change(screen.getByLabelText("Password input field"), { target: { value: "12345678" } });

        // Submit the form
        fireEvent.click(screen.getByLabelText("Submit to create a new account"));

        // Wait for the feedback message to appear
        await waitFor(() => {
            expect(screen.getByText("Account is made successfully")).toBeInTheDocument();
        });
    });

    it("shows password validation error", () => {
        render(
            <MemoryRouter>
                <LoginOrRegister login={false} />
            </MemoryRouter>,
        );

        // Fill out the form with an invalid password
        fireEvent.change(screen.getByLabelText("Name input field"), { target: { value: "John Doe" } });
        fireEvent.change(screen.getByLabelText("Email input field"), { target: { value: "john@example.com" } });
        fireEvent.change(screen.getByLabelText("Password input field"), { target: { value: "123" } }); // too short

        // Submit the form
        fireEvent.click(screen.getByLabelText("Submit to create a new account"));

        // Check if the feedback message updates to "Password is not valid"
        expect(screen.getByText(/Password is not valid.*8 characters/i)).toBeInTheDocument();
    });
});

// Mock localStorage
beforeEach(() => {
    const mockLocalStorage = (() => {
        let store: { [key: string]: string } = {};
        return {
            getItem: (key: string) => store[key] || null,
            setItem: (key: string, value: string) => {
                store[key] = value.toString();
            },
            clear: () => {
                store = {};
            },
        };
    })();
    Object.defineProperty(window, "localStorage", { value: mockLocalStorage });
});

describe("Login functionality tests", () => {
    test("renders the login form correctly", () => {
        render(<LoginOrRegister login={true} />);

        // Check if the title and input fields are rendered
        expect(screen.getByLabelText("Login page")).toBeInTheDocument();
        expect(screen.getByLabelText("Enter account details")).toBeInTheDocument();
        expect(screen.getByPlaceholderText("Email")).toBeInTheDocument();
        expect(screen.getByPlaceholderText("Password")).toBeInTheDocument();
        expect(screen.getByLabelText("Login button")).toBeInTheDocument();
    });

    test("displays 'No such email' when logging in with non-existent email", () => {
        render(<LoginOrRegister login={true} />);

        // Simulate entering an email and password
        fireEvent.change(screen.getByPlaceholderText("Email"), { target: { value: "nonexistent@example.com" } });
        fireEvent.change(screen.getByPlaceholderText("Password"), { target: { value: "password123" } });

        // Simulate clicking the login button
        fireEvent.click(screen.getByLabelText("Login button"));

        // Check if the error message is displayed
        expect(screen.getByLabelText("No such email")).toBeInTheDocument();
    });

    test("displays 'Wrong password' when email is correct but password is wrong", () => {
        // Mock a user in localStorage
        const mockUser = [{ name: "Test User", email: "testuser@example.com", password: "correctpassword" }];
        window.localStorage.setItem("userProfiles", JSON.stringify(mockUser));

        render(<LoginOrRegister login={true} />);

        // Simulate entering the correct email but wrong password
        fireEvent.change(screen.getByPlaceholderText("Email"), { target: { value: "testuser@example.com" } });
        fireEvent.change(screen.getByPlaceholderText("Password"), { target: { value: "wrongpassword" } });

        // Simulate clicking the login button
        fireEvent.click(screen.getByLabelText("Login button"));

        // Check if the wrong password message is displayed
        expect(screen.getByLabelText("Wrong password")).toBeInTheDocument();
    });

    test("displays 'Log in successful' when email and password are correct", () => {
        // Mock a user in localStorage
        const mockUser = [{ name: "Test User", email: "testuser@example.com", password: "correctpassword" }];
        window.localStorage.setItem("userProfiles", JSON.stringify(mockUser));

        render(<LoginOrRegister login={true} />);

        // Simulate entering the correct email and password
        fireEvent.change(screen.getByPlaceholderText("Email"), { target: { value: "testuser@example.com" } });
        fireEvent.change(screen.getByPlaceholderText("Password"), { target: { value: "correctpassword" } });

        // Simulate clicking the login button
        fireEvent.click(screen.getByLabelText("Login button"));

        // Check if the success message is displayed
        expect(screen.getByLabelText("Log in successful")).toBeInTheDocument();
    });

    test("clears the form after successful login", () => {
        // Mock a user in localStorage
        const mockUser = [{ name: "Test User", email: "testuser@example.com", password: "correctpassword" }];
        window.localStorage.setItem("userProfiles", JSON.stringify(mockUser));

        render(<LoginOrRegister login={true} />);

        // Simulate entering the correct email and password
        const emailInput = screen.getByPlaceholderText("Email");
        const passwordInput = screen.getByPlaceholderText("Password");

        fireEvent.change(emailInput, { target: { value: "testuser@example.com" } });
        fireEvent.change(passwordInput, { target: { value: "correctpassword" } });

        // Simulate clicking the login button
        fireEvent.click(screen.getByLabelText("Login button"));

        // Check if the inputs are cleared
        expect(emailInput).toHaveValue("");
        expect(passwordInput).toHaveValue("");
    });
});
