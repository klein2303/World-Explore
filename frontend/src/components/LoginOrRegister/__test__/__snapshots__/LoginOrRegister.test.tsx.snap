// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginOrRegister Component - Login Tests > renders the login form correctly 1`] = `
<DocumentFragment>
  <section
    aria-label="Login og register component"
  >
    <section
      aria-label="Login"
      class="_loginOrRegister_d1f9f8"
    >
      <section
        aria-label="Title"
        class="_title_d1f9f8"
      >
        <h3
          aria-label="Login page"
        >
          Log in
        </h3>
        <h4
          aria-label="Enter account details"
        >
          Enter your WorldExplore account details
        </h4>
      </section>
      <hr
        class="_line_d1f9f8"
      />
      <section
        aria-label="Input fields"
        class="_inputSections_d1f9f8"
      >
        <input
          aria-label="Name or email input field for login"
          class="_inputField_d1f9f8"
          name="nameOrEmail"
          placeholder="Email or username"
          type="text"
          value=""
        />
        <input
          aria-label="Password input field for login"
          class="_inputField_d1f9f8"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
        <h5
          aria-label=""
        >
           
        </h5>
        <button
          aria-label="Login button"
          class="_submitButton_d1f9f8"
        >
          Log in
        </button>
      </section>
    </section>
  </section>
</DocumentFragment>
`;

exports[`LoginOrRegister Component - Register tests > renders the registration form correctly 1`] = `
<DocumentFragment>
  <section
    aria-label="Login og register component"
  >
    <section
      aria-label="Register a new account"
      class="_loginOrRegister_d1f9f8"
    >
      <section
        aria-label="Title"
        class="_title_d1f9f8"
      >
        <h3
          aria-label="Create a new account"
        >
          Create an account
        </h3>
      </section>
      <hr
        class="_line_d1f9f8"
      />
      <section
        aria-label="Input fields"
        class="_inputSections_d1f9f8"
      >
        <input
          aria-label="Name input field"
          class="_inputField_d1f9f8"
          name="name"
          placeholder="Please enter your name"
          type="text"
          value=""
        />
        <input
          aria-label="Email input field"
          class="_inputField_d1f9f8"
          name="email"
          placeholder="Please enter your email"
          type="email"
          value=""
        />
        <input
          aria-label="Password input field"
          class="_inputField_d1f9f8"
          name="password"
          placeholder="Please enter a password"
          type="password"
          value=""
        />
        <h5
          aria-label="Password must at least be 8 characters"
        >
           Password must at least be 8 characters
        </h5>
        <button
          aria-label="Submit to create a new account"
          class="_submitButton_d1f9f8"
        >
          Create account
        </button>
      </section>
    </section>
  </section>
</DocumentFragment>
`;

exports[`LoginOrRegister Component Tests > renders the registration form correctly 1`] = `
<DocumentFragment>
  <section
    aria-label="Login og register component"
  >
    <section
      aria-label="Register a new account"
      class="_loginOrRegister_d1f9f8"
    >
      <div
        aria-label="Title"
        class="_title_d1f9f8"
      >
        <h3
          aria-label="Create a new account"
        >
          Create an account
        </h3>
      </div>
      <hr
        class="_line_d1f9f8"
      />
      <section
        aria-label="Input fields"
        class="_inputSections_d1f9f8"
      >
        <input
          aria-label="Name input field"
          class="_inputField_d1f9f8"
          name="name"
          placeholder="Place enter your name"
          type="text"
          value=""
        />
        <input
          aria-label="Email input field"
          class="_inputField_d1f9f8"
          name="email"
          placeholder="Please enter your email"
          type="email"
          value=""
        />
        <input
          aria-label="Password input field"
          class="_inputField_d1f9f8"
          name="password"
          placeholder="Please enter a password"
          type="password"
          value=""
        />
        <h5
          aria-label="Password must at least be 8 characters"
        >
           Password must at least be 8 characters
        </h5>
        <button
          aria-label="Submit to create a new account"
          class="_submitButton_d1f9f8"
        >
          Create account
        </button>
      </section>
    </section>
  </section>
</DocumentFragment>
`;

exports[`LoginOrRegister Component Tests - register functionality > renders the registration form correctly 1`] = `
<DocumentFragment>
  <section
    aria-label="Login og register component"
  >
    <section
      aria-label="Register a new account"
      class="_loginOrRegister_d1f9f8"
    >
      <section
        aria-label="Title"
        class="_title_d1f9f8"
      >
        <h3
          aria-label="Create a new account"
        >
          Create an account
        </h3>
      </section>
      <hr
        class="_line_d1f9f8"
      />
      <section
        aria-label="Input fields"
        class="_inputSections_d1f9f8"
      >
        <input
          aria-label="Name input field"
          class="_inputField_d1f9f8"
          name="name"
          placeholder="Please enter your name"
          type="text"
          value=""
        />
        <input
          aria-label="Email input field"
          class="_inputField_d1f9f8"
          name="email"
          placeholder="Please enter your email"
          type="email"
          value=""
        />
        <input
          aria-label="Password input field"
          class="_inputField_d1f9f8"
          name="password"
          placeholder="Please enter a password"
          type="password"
          value=""
        />
        <h5
          aria-label="Password must at least be 8 characters"
        >
           Password must at least be 8 characters
        </h5>
        <button
          aria-label="Submit to create a new account"
          class="_submitButton_d1f9f8"
        >
          Create account
        </button>
      </section>
    </section>
  </section>
</DocumentFragment>
`;
